File: csharp/Leet/Today3/BinaryTreeMaximumPathSum.cs
// https://leetcode.com/problems/binary-tree-maximum-path-sum/
using Leet.Tree;
namespace Today3.BinaryTreeMaximumPathSum;

public class Solution
{
    public int MaxPathSum(TreeNode root)
    {
        int result = root.val;
        int DFS(TreeNode root)
        {
            if (root == null)
                return 0;

            if (root.left == null && root.right == null)
            {
                result = Math.Max(result, root.val);
                return root.val;
            }


            var leftMax = DFS(root.left);
            var rightMax = DFS(root.right);

            var withSplit = root.val + leftMax + rightMax;
            var withoutSplit = root.val + Math.Max(Math.Max(leftMax, rightMax), 0);

            result = Math.Max(result, withSplit);
            result = Math.Max(result, withoutSplit);

            return withoutSplit;
        }
        DFS(root);
        return result;
    }
}

File: returned/python/today10/binary_tree_maximum_path_sum.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = float('-inf')
        
        def dfs(root: Optional[TreeNode]) -> int:
            if not root:
                return 0
            
            leftMax = dfs(root.left)
            rightMax = dfs(root.right)
            leftMax = max(leftMax, 0)
            rightMax = max(rightMax, 0)

            nonlocal res
            res = max(res, root.val + leftMax + rightMax)
            
            return root.val + max(leftMax, rightMax)

        dfs(root)
        return res
